subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/AbstractVirtualFileSystem.java

```diff
@@ -45,31 +48,57 @@ public Optional<MetadataSnapshot> findMetadata(String absolutePath) {
     }
 
     @Override
-    public void store(String absolutePath, FileSystemLocationSnapshot snapshot) {
-        rootReference.update(root -> updateNotifyingListeners(diffListener -> root.store(absolutePath, snapshot, diffListener)));
+    public FileSystemLocationSnapshot store(String absolutePath, Supplier<FileSystemLocationSnapshot> snapshotSupplier) {
+        long versionBefore = versionHierarchyRoot.getVersionFor(absolutePath);
+        FileSystemLocationSnapshot snapshot = snapshotSupplier.get();
+        updateCheckingVersion(absolutePath, versionBefore, snapshot);
+        return snapshot;
+    }
+
+    @Override
+    public FileSystemLocationSnapshot store(String baseLocation, StoringAction storingAction) {
+        long versionBefore = versionHierarchyRoot.getVersionFor(baseLocation);
+        return storingAction.snapshot(snapshot -> updateCheckingVersion(snapshot.getAbsolutePath(), versionBefore, snapshot));
+    }
+
+    private void updateCheckingVersion(String absolutePath, long versionBefore, FileSystemLocationSnapshot snapshot) {
+        long versionAfter = versionHierarchyRoot.getVersionFor(absolutePath);
+        // Only update VFS if no changes happened in between
+        // The version in sub-locations may be smaller than the version we queried at the root when using a `StoringAction`.
+        if (versionBefore >= versionAfter) {
+            rootReference.updateUnderLock(root -> updateNotifyingListeners(diffListener -> root.store(absolutePath, snapshot, diffListener)));
+        } else {
+            LOGGER.debug("Changes to the virtual file system happened while snapshotting '{}', not storing resulting snapshot", absolutePath);
+        }
     }
 
     @Override
     public void invalidate(Iterable<String> locations) {
         LOGGER.debug("Invalidating VFS paths: {}", locations);
-        rootReference.update(root -> {
+        rootReference.updateUnderLock(root -> {
             SnapshotHierarchy result = root;
+            VersionHierarchyRoot newVersionHierarchyRoot = versionHierarchyRoot;
             for (String location : locations) {
                 SnapshotHierarchy currentRoot = result;
                 result = updateNotifyingListeners(diffListener -> currentRoot.invalidate(location, diffListener));
+                newVersionHierarchyRoot = newVersionHierarchyRoot.increaseVersion(location);
```

Comment:
I still don't like the name `increaseVersion()` here. We are making the location be the latest there is in the tree, not increasing its current version. I think a name like `touch()` or even `updateVersion()` would be better.

Reply:
I am using `touch()` if there isn't a version specified and `updateVersion()` if there is.